Descriptive error for missing git tags

@jacobtkeio was hitting this problem in the CI for his fork last week, and it took me an hour and a half to figure out why. It turns out, since his fork didn't have tags, running `git push --tags` made the whole world happy again.

This is a problem in flux-core too since both repositories determine version using `git describe`.

But, since this has come up a few times, I'm going to look in to making a better error message:

```
git fetch tags
  fatal: --unshallow on a complete repository does not make sense
build setup
autogen.sh
/usr/src
configure --prefix=/usr --sysconfdir=/etc --with-systemdsystemunitdir=/etc/systemd/system \
    --localstatedir=/var --prefix=/usr --sysconfdir=/etc
  Re-run cmake no build system arguments
  -- VER 0cb6cde5
  CMake Error at CMakeLists.txt:15 (project):
    VERSION "0cb6cde5" format invalid.
   ```

It's a different problem in core. Core will build with an invalid version, but fails on `make check`

```
In file included from version.c:15:
version.c: In function ‘flux_core_version’:
version.h:25:37: error: ‘ca5949d’ undeclared (first use in this function)
   25 | #define FLUX_CORE_VERSION_MAJOR     ca5949d
      |                                     ^~~~~~~
version.c:25:18: note: in expansion of macro ‘FLUX_CORE_VERSION_MAJOR’
   25 |         *major = FLUX_CORE_VERSION_MAJOR;
      |                  ^~~~~~~~~~~~~~~~~~~~~~~
version.h:25:37: note: each undeclared identifier is reported only once for each function it appears in
   25 | #define FLUX_CORE_VERSION_MAJOR     ca5949d
      |                                     ^~~~~~~
version.c:25:18: note: in expansion of macro ‘FLUX_CORE_VERSION_MAJOR’
   25 |         *major = FLUX_CORE_VERSION_MAJOR;
      |                  ^~~~~~~~~~~~~~~~~~~~~~~
version.h:26:37: error: ‘d’ undeclared (first use in this function)
   26 | #define FLUX_CORE_VERSION_MINOR     d
      |                                     ^
version.c:27:18: note: in expansion of macro ‘FLUX_CORE_VERSION_MINOR’
   27 |         *minor = FLUX_CORE_VERSION_MINOR;
      |                  ^~~~~~~~~~~~~~~~~~~~~~~
version.c:29:41: error: expected expression before ‘;’ token
   29 |         *patch = FLUX_CORE_VERSION_PATCH;
      |                                         ^
version.h:33:63: error: expected expression before ‘<<’ token
   33 |                                      (FLUX_CORE_VERSION_PATCH << 0))
      |                                                               ^~
version.c:30:12: note: in expansion of macro ‘FLUX_CORE_VERSION_HEX’
   30 |     return FLUX_CORE_VERSION_HEX;
      |            ^~~~~~~~~~~~~~~~~~~~~
version.c:31:1: error: control reaches end of non-void function [-Werror=return-type]
   31 | }
      | ^
cc1: all warnings being treated as errors
```
Autoconf takes the version of a shallow clone (depth 1) as:
```
checking Major version... ca5949d
checking Minor version... d
checking Point version...
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -12,6 +12,28 @@ else()
 endif()
 message(STATUS "VER ${FLUX_SCHED_VER}")
 string(REGEX REPLACE "-.*$" "" FLUX_SCHED_VER_NOGIT "${FLUX_SCHED_VER}")
+# I took this regex from the CMake source code. If the version fails to
+# match it, CMake will error out. We can help the end-user get past this
+# with a few tips.
+string(REGEX MATCH "(^([0-9]+(\.[0-9]+(\.[0-9]+(\.[0-9]+)?)?)?)?$)" CHECK_FLUX_SCHED_VER "${FLUX_SCHED_VER_NOGIT}")
+string(LENGTH "${CHECK_FLUX_SCHED_VER}" LENGTH_FLUX_SCHED_VER)
+if (LENGTH_FLUX_SCHED_VER STREQUAL "0")
+    message(WARNING 
+    "CMake may generate an error that says \"VERSION \"${FLUX_SCHED_VER}\" format invalid.\"\n"
+    "If this happens, try the following:\n"
+    "    1. Set the variable manually, with `cmake -DFLUX_SCHED_VER=<version> ...`\n"
+    "       or FLUX_SCHED_VERSION=<version> in your environment.\n"
+    "    2. If you are running in a CI environment, run `git fetch --tags`\n"
+    "       before building. Versions in flux-sched are derived from \n"
+    "       `git describe` which uses the most recent tag.\n"
+    "    3. If you are running in a fork of the main repository, try\n"
+    "       `git push --tags` to make sure tags are synchronized in \n"
+    "       your fork.\n"
+    "    4. If you are running outside of a repo (such as in a buildfarm),\n"
+    "       add a file to the source directory called flux-sched.ver and\n"
+    "       place a valid version string in that file."
+    )
+endif()
 project(flux-sched VERSION ${FLUX_SCHED_VER_NOGIT} LANGUAGES CXX C)
 message(STATUS "Building flux-sched version ${FLUX_SCHED_VER}")
 

